pgrx/src/callconv.rs

```diff
@@ -283,3 +283,182 @@ where
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.fncollation
+    }
+
+    /// Retrieve the type (as an Oid) of argument number `num`.
+    /// In other words, the type of `self.raw_args()[num]`
+    #[inline]
+    pub fn pg_getarg_type(&self, num: usize) -> pg_sys::Oid {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            pg_sys::get_fn_expr_argtype(self.0.as_ref().unwrap().flinfo, num as std::os::raw::c_int)
+        }
+    }
+
+    /// Retrieve the `.flinfo.fn_extra` pointer (as a PgBox'd type) from [`pg_sys::FunctionCallInfo`].
+    pub fn pg_func_extra<ReturnType, DefaultValue: FnOnce() -> ReturnType>(
+        &self,
+        default: DefaultValue,
+    ) -> PgBox<ReturnType> {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            let fcinfo = PgBox::from_pg(self.0);
+            let mut flinfo = PgBox::from_pg(fcinfo.flinfo);
+            if flinfo.fn_extra.is_null() {
+                flinfo.fn_extra = PgMemoryContexts::For(flinfo.fn_mcxt)
+                    .leak_and_drop_on_delete(default())
+                    as crate::void_mut_ptr;
+            }
+
+            PgBox::from_pg(flinfo.fn_extra as *mut ReturnType)
+        }
+    }
+
+    #[inline]
+    pub fn srf_is_first_call(&self) -> bool {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe { (*(*self.0).flinfo).fn_extra.is_null() }
+    }
+
+    /// Thin wrapper around [`pg_sys::init_MultiFuncCall`], made necessary
+    /// because this structure's FunctionCallInfo is a private field.
+    #[inline]
+    pub unsafe fn init_multi_func_call(&mut self) -> FcContext<'fcx> {
+        unsafe { FcContext::assume_valid(pg_sys::init_MultiFuncCall(self.0)) }
+    }
+
+    /// Thin wrapper around [`pg_sys::per_MultiFuncCall`], made necessary
+    /// because this structure's FunctionCallInfo is a private field.
+    #[inline]
+    pub unsafe fn per_multi_func_call(&mut self) -> FcContext<'fcx> {
+        unsafe { FcContext::assume_valid(pg_sys::per_MultiFuncCall(self.0)) }
+    }
+    /// Equivalent to "per_MultiFuncCall" with no FFI cost, and a lifetime
+    /// constraint.
+    ///
+    /// Safety: This is only equivalent to the current (as of Postgres 16.3)
+    /// implementation of [`pg_sys::per_MultiFuncCall`], and future changes
+    /// to the complexity of the calling convention may break this method.
+    pub(crate) unsafe fn deref_fcx(&mut self) -> FcContext<'fcx> {
+        unsafe { FcContext::assume_valid((*(*self.0).flinfo).fn_extra.cast()) }
```

Comment:
They cannot really change this without breaking basically all SRFs ever. They only got away with the FunctionCallInfoBaseData change because the primary entity that constructs FunctionCallInfo to call functions with is Postgres itself.

Reply:
I removed that safety warning and replaced it with one that simply states `self.0.flinfo.fn_extra` must be non-null